crates/storage/nippy-jar/src/cursor.rs

```diff
@@ -0,0 +1,205 @@
+            if filter.contains(key)? {
+                // May have false positives
+                if let Some(row_index) = phf.get_index(key)? {
+                    self.row =
+                        self.jar.offsets_index.access(row_index as usize).expect("built from same")
+                            as u64;
+                    return self.next_row()
+                }
+            }
+        }
+
+        Err(NippyJarError::UnsupportedFilterQuery)
+    }
+
+    /// Returns a row by its number.
+    pub fn row_by_number(&mut self, row: usize) -> Result<Option<Row>, NippyJarError> {
+        self.row = row as u64;
+        self.next_row()
+    }
+
+    /// Returns the current value and advances the row.
+    pub fn next_row(&mut self) -> Result<Option<Row>, NippyJarError> {
+        if self.row as usize * self.jar.columns >= self.jar.offsets.len() {
+            // Has reached the end
+            return Ok(None)
+        }
+
+        let mut row = Vec::with_capacity(self.jar.columns);
+
+        // Retrieve all column values from the row
+        for column in 0..self.jar.columns {
+            self.read_value(column, &mut row)?;
+        }
+
+        self.row += 1;
+
+        Ok(Some(row))
+    }
+
+    /// Returns a row, searching it by a key used during [`NippyJar::prepare_index`]  by using a
+    /// `MASK` to only read certain columns from the row.
+    ///
+    /// **May return false positives.**
+    ///
+    /// Example usage would be querying a transactions file with a transaction hash which is **NOT**
+    /// stored in file.
+    pub fn row_by_key_with_cols<const MASK: usize, const COLUMNS: usize>(
+        &mut self,
+        key: &[u8],
+    ) -> Result<Option<Row>, NippyJarError> {
+        if let (Some(filter), Some(phf)) = (&self.jar.filter, &self.jar.phf) {
+            // TODO: is it worth to parallize both?
+
+            // May have false positives
+            if filter.contains(key)? {
+                // May have false positives
+                if let Some(row_index) = phf.get_index(key)? {
+                    self.row =
+                        self.jar.offsets_index.access(row_index as usize).expect("built from same")
+                            as u64;
+                    return self.next_row_with_cols::<MASK, COLUMNS>()
+                }
+            }
+        }
+
+        Err(NippyJarError::UnsupportedFilterQuery)
+    }
+
+    /// Returns a row by its number by using a `MASK` to only read certain columns from the row.
+    pub fn row_by_number_with_cols<const MASK: usize, const COLUMNS: usize>(
+        &mut self,
+        row: usize,
+    ) -> Result<Option<Row>, NippyJarError> {
+        self.row = row as u64;
+        self.next_row_with_cols::<MASK, COLUMNS>()
+    }
+
+    /// Returns the current value and advances the row.
+    ///
+    /// Uses a `MASK` to only read certain columns from the row.
+    pub fn next_row_with_cols<const MASK: usize, const COLUMNS: usize>(
```

Comment:
how many different combinations are we expecting? this could generate a lot of code 

and I assume the benefit of having this const is neglible?